rackspace-infrastructure-automation/aws-terraform-vpc_basenetwork
# Add new vars to provide map of tags to be applied to pub OR priv subnets

**Please note:** _these changes are required for a build expecting to start 2019-02-12._

##### Corresponding Issue(s) or trello card(s):
Originally raised by @srirajan in the context of CloudFormation, with @gdelvalle commenting that it should be considered for Terraform as well: https://github.com/rax-brazil/TroposphereTemplates/issues/1486

##### Summary of change(s):
Add additional vars to allow for tags specifically for public or private subnets. Primary use case is context of needing to provide tags that Kubernetes can identify for resource management.

Given the additional info @gdelvalle provided on the opening issue, the solve has been to use a list for the two new variables that you then provide a number of maps based on your subnet requirements. If the list is length of 1 (defaults to `[{}]`) then all subnets in the private or public half get the tags; if however you are building multiple subnets per AZ and provide a list of an equivalent length, you can tag one set of subnets with one thing, and another set with another or no things.

##### Will the change trigger resource destruction or replacement? If yes, please provide justification:
No

##### Does this update/change involve issues with other external modules? If so, please describe the scenario.
No

##### If input variables or output variables have changed or has been added, have you updated the README?
Yes

##### Do examples need to be updated based on changes?
EKS could be updated, not in scope of this PR.

##### Testing

1) built with first approach of static map applied to all subnets
  - applies to all subnets
2) using new module with list of maps results in no op
3) change subnets per AZ to 2 with additional CIDRs results in plan for new subnets providing only one map in tag lists
  - all subnets are tagged with the provided tags in the single map
4) change local to include an empty second map in the list so second subnets have no k8s role tag

```
------------------------------------------------------------------------

An execution plan has been generated and is shown below.
Resource actions are indicated with the following symbols:
  ~ update in-place

Terraform will perform the following actions:

  ~ module.vpc.aws_subnet.private_subnet[2]
      tags.%:                               "5" => "4"
      tags.kubernetes.io/role/internal-elb: "1" => ""

  ~ module.vpc.aws_subnet.private_subnet[3]
      tags.%:                               "5" => "4"
      tags.kubernetes.io/role/internal-elb: "1" => ""

  ~ module.vpc.aws_subnet.public_subnet[2]
      tags.%:                               "5" => "4"
      tags.kubernetes.io/role/internal-elb: "1" => ""

  ~ module.vpc.aws_subnet.public_subnet[3]
      tags.%:                               "5" => "4"
      tags.kubernetes.io/role/internal-elb: "1" => ""


Plan: 0 to add, 4 to change, 0 to destroy.

------------------------------------------------------------------------
```

##### Note to the PR requester about Closing PR's
Please message the person that opened the issue when auto closing it on slack, as well as any other stake holders of deep interest. Only close the issue if you believe that the issue is fully resolved with this PR.

#### This PR may auto close the issue associated with it. If you feel the issue is not resolved please reopen the issue.
